json_encoder.go

```diff
@@ -158,37 +177,37 @@ func (enc *jsonEncoder) AddFields(fields []Field) {
 // the encoder's accumulated fields. It doesn't modify or lock the encoder's
 // underlying byte slice. It's safe to call from multiple goroutines, but it's
 // not safe to call WriteEntry while adding fields.
-func (enc *jsonEncoder) WriteEntry(sink io.Writer, e *Entry) error {
+func (enc *jsonEncoder) WriteEntry(sink io.Writer, msg string, lvl Level, t time.Time) error {
 	if sink == nil {
 		return errNilSink
 	}
-	if e == nil {
-		return errNilEntry
-	}
-	// Grab an encoder from the pool so that we can re-use the underlying
-	// buffer.
-	final := newJSONEncoder()
-	defer final.Free()
-
-	final.bytes = append(final.bytes, `{"msg":"`...)
-	final.safeAddString(e.Message)
-	final.bytes = append(final.bytes, `","level":"`...)
-	final.bytes = append(final.bytes, e.Level.String()...)
-	final.bytes = append(final.bytes, `","ts":`...)
-	final.bytes = strconv.AppendFloat(final.bytes, timeToSeconds(e.Time), 'f', -1, 64)
+
+	final := enc.Clone().(*jsonEncoder)
+
+	// TODO: When we flatten the serialized messages, remove the extra truncate
+	// and copy below. This will change the shape of the serialized message, which
+	// will require updating many tests.
+	final.truncate()
+	final.bytes = append(final.bytes, '{')
+	final.AddFields([]Field{
```

Comment:
Rather than build up a slice, and then iterate over the slice etc, it will probably be faster to do:
```go
final.messageF(msg).AddTo(final)
final.levelF(lvl).AddTo(final)
final.timeF(t).AddTo(final)
```

Avoids a bunch of indirection.